csds-client/client/client.go

```diff
@@ -0,0 +1,228 @@
+func (c *Client) connWithAuth() error {
+	var scope string
+	switch c.info.authnMode {
+	case "jwt":
+		if c.info.jwt == "" {
+			return errors.New("missing jwt file")
+		}
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.info.jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return nil
+		}
+	case "auto":
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.info.uri {
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := getValueByKeyFromNodeMatcher(c.nm, key); projectNum != "" {
+				c.md = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return errors.New("auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return errors.New("invalid authn_mode")
+	}
+}
+
+// New creates a new client
+func New() (*Client, error) {
+	c := &Client{
+		info: ParseFlags(),
+	}
+	if c.info.platform != "gcp" {
+		return nil, fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.info.platform)
+	}
+	if c.info.apiVersion != "v2" {
+		return nil, fmt.Errorf("%s api version is not supported, list of supported api versions: v2", c.info.apiVersion)
+	}
+
+	if err := c.parseNodeMatcher(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run connects the client to the uri and calls doRequest
+func (c *Client) Run() error {
+	if err := c.connWithAuth(); err != nil {
+		return err
+	}
+	defer c.cc.Close()
+
+	c.csdsClient = csdspb.NewClientStatusDiscoveryServiceClient(c.cc)
+	var ctx context.Context
+	if c.md != nil {
+		ctx = metadata.NewOutgoingContext(context.Background(), c.md)
+	} else {
+		ctx = context.Background()
+	}
+	streamClientStatus, err := c.csdsClient.StreamClientStatus(ctx)
+	if err != nil {
+		return err
+	}
+
+	// run once or run with monitor mode
+	for {
+		if err := c.doRequest(streamClientStatus); err != nil {
+			return err
+		}
+		if c.info.monitorInterval != 0 {
+			time.Sleep(c.info.monitorInterval)
```

Comment:
This might be out of scope or unintended behavior but if we are running in watch mode would it make sense to pick up changes to the config file?

Reply:
May I add this feature to next visualization PR or refactor PR so that this PR could be easier to review and merge soon?